File: GAIL/Audio.cpp
#include "GAIL.hpp"
#include <iostream>
#include <fstream>

namespace GAIL {
    int toInt(char* buffer, std::size_t len) {
        int a = 0;
        std::memcpy(&a, buffer, len); // little endian
        return a;
    };
    Sound SoundEffect::Use(Sound baseSound) {
        return baseSound;
    };
    Sound::Sound(int sampleRate, std::vector<char> rawData, SoundFormat format, std::vector<SoundEffect> soundEffects) : sampleRate{sampleRate}, format{format}, rawData{rawData}, soundEffects{soundEffects} {
        alGenBuffers(1, &this->buffer);

        Sound sound = *this;
        for (int i = 0; i < this->soundEffects.size(); i++)
        {
            sound = this->soundEffects[i].Use(sound);
        };
        
        alBufferData(this->buffer, sound.format, sound.rawData.data(), sound.rawData.size(), sound.sampleRate);
        
        delete &sound;

        alGenSources(1, &source);
        alSourcef(source, AL_PITCH, 1);
        alSourcef(source, AL_GAIN, 1);
        alSourcei(source, AL_LOOPING, AL_FALSE);
        alSourcei(source, AL_BUFFER, buffer);
    };
    Sound::~Sound() {
        alDeleteSources(1, &this->source);
        alDeleteBuffers(1, &this->buffer);
        this->soundEffects.clear();
        this->rawData.clear();
        delete &this->rawData;
        delete &this->soundEffects;
    };
    void Sound::Update() {
        alDeleteSources(1, &this->source);
        alDeleteBuffers(1, &this->buffer);

        alGenBuffers(1, &this->buffer);

        Sound sound = *this;
        for (int i = 0; i < this->soundEffects.size(); i++)
        {
            sound = this->soundEffects[i].Use(sound);
        };
        
        alBufferData(this->buffer, sound.format, sound.rawData.data(), sound.rawData.size(), sound.sampleRate);
        
        delete &sound;

        alGenSources(1, &source);
        alSourcef(source, AL_PITCH, this->pitch);
        alSourcef(source, AL_GAIN, this->volume);
        alSourcei(source, AL_LOOPING, isLooping ? AL_TRUE : AL_FALSE);
        alSourcei(source, AL_BUFFER, buffer);
    };
    void Sound::SetSoundEffects(std::vector<SoundEffect> soundEffects) { this->soundEffects = soundEffects; };

    Sound Sound::FromWAV(string path) {
        std::ifstream file(path, std::ios::binary);
        if(!file.is_open()) {
            std::cerr << "GAIL: FromWAV (load file): Could not open \"" << path << "\"" << std::endl;
            return Sound();
        }
        char buffer[4];
        // RIFF
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read RIFF" << std::endl;
            return Sound();
        }
        if(std::strncmp(buffer, "RIFF", 4) != 0) {
            std::cerr << "GAIL: FromWAV (load file): file is not a valid WAVE file (header doesn't begin with RIFF)" << std::endl;
            return Sound();
        }
        // ChunkSize of RIFF
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read ChunkSize" << std::endl;
            return Sound();
        }
        // format
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Format" << std::endl;
            return Sound();
        }
        if(std::strncmp(buffer, "WAVE", 4) != 0) {
            std::cerr << "GAIL: FromWAV (load file): file is not a valid WAVE file (header doesn't contain WAVE)" << std::endl;
            return Sound();
        }
        // "fmt/0"
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Subchunk1ID" << std::endl;
            return Sound();
        }
        // 16, for PCM.
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Subchunk1Size" << std::endl;
            return Sound();
        }
        // Audio format (1=PCM)
        if(!file.read(buffer, 2)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Audio format" << std::endl;
            return Sound();
        }
        // NumChannels
        if(!file.read(buffer, 2)) {
            std::cerr << "GAIL: FromWAV (load file): could not read NumChannels" << std::endl;
            return Sound();
        }
        int channels = toInt(buffer, 2);

        // sample rate
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read sample rate" << std::endl;
            return Sound();
        }
        int sampleRate = toInt(buffer, 4);
        // (sampleRate * bitsPerSample * channels) / 8
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read (sampleRate * bitsPerSample * channels) / 8" << std::endl;
            return Sound();
        }
        // channels * BitsPerSample/8
        if(!file.read(buffer, 2)) {
            std::cerr << "GAIL: FromWAV (load file): could not read BlockAlign" << std::endl;
            return Sound();
        }
        // bitsPerSample
        if(!file.read(buffer, 2)) {
            std::cerr << "GAIL: FromWAV (load file): could not read bits per sample" << std::endl;
            return Sound();
        }
        int bitsPerSample = toInt(buffer, 2);
        // Subchunk2ID (data)
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Subchunk2ID" << std::endl;
            return Sound();
        }
        if(std::strncmp(buffer, "data", 4) != 0) {
            std::cerr << "GAIL: FromWAV (load file): file is not a valid WAVE file (doesn't have 'data' tag)" << std::endl;
            return Sound();
        }
        // Subchunk2Size
        if(!file.read(buffer, 4)) {
            std::cerr << "GAIL: FromWAV (load file): could not read Subchunk2Size" << std::endl;
            return Sound();
        }
        int size = toInt(buffer, 4);
        /* cannot be at the end of file */
        if(file.eof()) {
            std::cerr << "GAIL: FromWAV (load file): No data." << std::endl;
            return Sound();
        }
        if(file.fail()) {
            std::cerr << "GAIL: FromWAV (load file): fail state set on the file" << std::endl;
            return Sound();
        }

        std::vector<char> data = std::vector<char>(size);
        file.read(data.data(), size);
        file.close();
        return Sound(sampleRate, data, channels==1&&bitsPerSample==8?SoundFormat::Mono8:channels==2&&bitsPerSample==8?SoundFormat::Stereo8:channels==1&&bitsPerSample==16?SoundFormat::Mono16:channels==2&&bitsPerSample==16?SoundFormat::Stereo16:SoundFormat::Stereo16);
    }
    std::vector<string> SoundCapture::GetCaptureDevices() {
        std::vector<string> captureDevices;
        const ALCchar* devices = alcGetString(nullptr, ALC_CAPTURE_DEVICE_SPECIFIER);
        const char* ptr = devices;

        captureDevices.clear();

        do {
            captureDevices.push_back(std::string(ptr));
            ptr += captureDevices.back().size() + 1;
        } while(*(ptr + 1) != '\0');
        return captureDevices;
    };
    SoundCapture::SoundCapture(SoundFormat format, int sampleRate, string captureDevice) {
        this->captureDevice = alcCaptureOpenDevice(captureDevice.c_str(), sampleRate, format, sampleRate);
        this->sampleRate = sampleRate;
        this->format = format;
    };
    SoundCapture::~SoundCapture() {
        alcCaptureCloseDevice(this->captureDevice);
    };
    void SoundCapture::Start() {
        alcCaptureStart(this->captureDevice);
    };
    Sound SoundCapture::Stop() {
        alcCaptureStop(captureDevice);
        ALCint samplesAvailable;
        alcGetIntegerv(captureDevice, ALC_CAPTURE_SAMPLES, 1, &samplesAvailable);
        std::vector<char> buffer;
        if (samplesAvailable > 0) {
            buffer.resize(this->format==SoundFormat::Mono16||this->format==SoundFormat::Stereo16 ? samplesAvailable*2 : samplesAvailable);

            alcCaptureSamples(captureDevice, buffer.data(), samplesAvailable);
        }
        return Sound(this->sampleRate, buffer, this->format);
    };
}   

File: GAIL/Audio.cs
using Silk.NET.OpenAL;
using Silk.NET.OpenAL.Extensions.EXT;

namespace GAIL
{
    /// <summary>
    /// A sound format: how many channels and the bandwidth.
    /// </summary>
    public enum SoundFormat {
        Stereo16 = BufferFormat.Stereo16,
        Mono16 = BufferFormat.Mono16,
        Stereo8 = BufferFormat.Stereo8,
        Mono8 = BufferFormat.Mono8
    }
    /// <summary>
    /// An effect template for sounds.
    /// </summary>
    public interface ISoundEffect {
        /// <summary>
        /// Applies effects to the sound.
        /// </summary>
        /// <param name="baseSound">The original sound.</param>
        /// <returns>The effected sound.</returns>
        Sound Use(Sound baseSound);
    }
    /// <summary>
    /// Represents a sound (audio) with effects.
    /// </summary>
    public class Sound : IDisposable {
        /// <param name="sampleRate"><example>example: 44100(Hz)</example></param>
        /// <param name="format">The format of the rawData.</param>
        /// <param name="rawData">Raw sound data is in PCM</param>
        public Sound(int sampleRate, SoundFormat format, List<byte> rawData) {
            AL al = AL.GetApi();
            
            buffer = al.GenBuffer();

            Sound sound = this;
            foreach (ISoundEffect effect in soundEffects) {
                sound = effect.Use(sound);
            }

            al.BufferData(buffer, (BufferFormat)format, rawData.ToArray(), sampleRate);
            source = al.GenSource();
            al.SetSourceProperty(source, SourceFloat.Pitch, pitch);
            al.SetSourceProperty(source, SourceFloat.Gain, volume);
            al.SetSourceProperty(source, SourceBoolean.Looping, isLooping);
            al.SetSourceProperty(source, SourceInteger.Buffer, buffer);
        }
        public void Dispose() {
            AL al = AL.GetApi();
            al.DeleteSource(source);
            al.DeleteBuffer(buffer);
            soundEffects.Clear();
            rawData.Clear();
        }

        /// <summary>
        /// The format of this sound.
        /// </summary>
        public readonly SoundFormat format;
        /// <summary>
        /// The duration of this sound (in seconds).
        /// </summary>
        public double Duration {get {
            return rawData.Count / (format == SoundFormat.Mono16||format == SoundFormat.Stereo16? 2 : 1) / sampleRate;
        }}
        /// <summary>
        /// The sample rate of this sound (in Hertz).
        /// </summary>
        public readonly int sampleRate;
        /// <summary>
        /// The raw sound data (when the bandwidth is 16 bits, 2 bytes per sample).
        /// </summary>
        public List<byte> rawData = [];
        /// <summary>
        // If this sound loops (default: false).
        /// </summary>
        public bool isLooping = false;
        /// <summary>
        /// The volume of this sound (default: 1).
        /// </summary>
        public float volume = 1;
        /// <summary>
        /// The pitch of this sound (default: 1).
        /// </summary>
        public float pitch = 1;
        /// <summary>
        /// The source (openAL), for custom usage.
        /// </summary>
        public uint source;
        /// <summary>
        /// The buffer (openAL), for custom usage.
        /// </summary>
        public uint buffer;

        /// <summary>
        /// All of the current sound effects.
        /// </summary>
        List<ISoundEffect> soundEffects = [];
        
        /// <summary>
        /// Updates the current source and buffer (call this when you use OpenAL).
        /// Also Applies all the sound effects (from first to last).
        /// </summary>
        public void Update() {
            AL al = AL.GetApi();

            al.DeleteSource(source);
            al.DeleteBuffer(buffer);
            
            buffer = al.GenBuffer();

            Sound sound = this;
            foreach (ISoundEffect effect in soundEffects) {
                sound = effect.Use(sound);
            }

            al.BufferData(buffer, (BufferFormat)format, rawData.ToArray(), sampleRate);
            source = al.GenSource();
            al.SetSourceProperty(source, SourceFloat.Pitch, pitch);
            al.SetSourceProperty(source, SourceFloat.Gain, volume);
            al.SetSourceProperty(source, SourceBoolean.Looping, isLooping);
            al.SetSourceProperty(source, SourceInteger.Buffer, buffer);
        }
        /// <summary>
        /// Sets the sound effects for later use.
        /// </summary>
        /// <param name="soundEffects">The sound effects</param>
        public void SetSoundEffects(List<ISoundEffect> soundEffects) { this.soundEffects = soundEffects; }
        /// <summary>
        /// Loads a sound from a wav file with a path to it.
        /// </summary>
        /// <param name="path">The path of the WAVE formatted file</param>
        /// <returns></returns>
        public static Sound FromWAV(string path) {
            using FileStream stream = new(path, FileMode.Open, FileAccess.Read);
            byte[] buffer = new byte[4];
            stream.Read(buffer, 0, 4);
            if (buffer != "RIFF".ToCharArray().Select(c => (byte)c)) {
                throw new FileFormatException(path, "RIFF header doesn't match");
            }
            stream.Read(buffer, 8, 4);
            if (buffer != "WAVE".ToCharArray().Select(c => (byte)c)) {
                throw new FileFormatException(path, "file format isn't WAVE");
            }
            stream.Read(buffer, 12, 4);
            if (buffer != "fmt ".ToCharArray().Select(c => (byte)c)) {
                throw new FileFormatException(path, "'fmt ' header doesn't match");
            }
            stream.Read(buffer, 22, 2); // num channels
            if (!BitConverter.IsLittleEndian)
                Array.Reverse(buffer);
            short channels = BitConverter.ToInt16(buffer, 0);
            stream.Read(buffer, 24, 4); // Sample rate
            if (!BitConverter.IsLittleEndian)
                Array.Reverse(buffer);
            int sampleRate = BitConverter.ToInt32(buffer, 0);
            stream.Read(buffer, 34, 2); // BitsPerSample
            if (!BitConverter.IsLittleEndian)
                Array.Reverse(buffer);
            short bitsPerSample = BitConverter.ToInt16(buffer, 0);

            stream.Read(buffer, 36, 4);
            if (buffer != "data".ToCharArray().Select(c => (byte)c)) {
                throw new FileFormatException(path, "data header doesn't match");
            }
            stream.Read(buffer, 0, 4);
            if (!BitConverter.IsLittleEndian)
                Array.Reverse(buffer);
            int size = BitConverter.ToInt32(buffer, 0);
            byte[] data = new byte[size]; 
            stream.Read(data, 44, size);

            return new(sampleRate, channels==1&&bitsPerSample==8?SoundFormat.Mono8:channels==2&&bitsPerSample==8?SoundFormat.Stereo8:channels==1&&bitsPerSample==16?SoundFormat.Mono16:channels==2&&bitsPerSample==16?SoundFormat.Stereo16:SoundFormat.Stereo16, [.. data]);
        }
        /// <summary>
        /// Reads audio from a microphone or sound output.
        /// </summary>
        public class SoundCapture : IDisposable {
            /// <summary>
            /// the sample rate of this sound capture.
            /// </summary>
            public readonly uint sampleRate;
            /// <summary>
            /// The format of this sound capture.
            /// </summary>
            public readonly SoundFormat format;
            /// <summary>
            /// The capture device, used for custom usage.
            /// </summary>
            public Pointer<Device> captureDevice;
            /// <summary>
            /// The openAL device, used for custom usage.
            /// </summary>
            public Pointer<Device> device;

            /// <summary>
            /// Gets all the capture devices' names.
            /// </summary>
            /// <returns>A list of all the capture devices.</returns>
            public static unsafe List<string> GetCaptureDevices() {
                return [.. ALContext.GetApi().GetContextProperty(null, GetContextString.DeviceSpecifier).Split('\0')];
            }
            /// <summary>
            /// Creates a new sound capture.
            /// </summary>
            /// <param name="format">The format in which to capture the sound.</param>
            /// <param name="sampleRate">The sample rate in which to capture the sound.</param>
            /// <param name="deviceName">Which capture device to use. (Get name at <see cref="GetCaptureDevices"/> <b>OR pass an empty string for the default</b>)</param>
            public unsafe SoundCapture(SoundFormat format, uint sampleRate, string deviceName) {
                ALContext alc = ALContext.GetApi();
                device = alc.OpenDevice(deviceName);
                Pointer<Context> context = alc.CreateContext(device, null);
                alc.MakeContextCurrent(context);
                if (alc.TryGetExtension(device, out Capture cap)) {
                    throw new APIBackendException("OpenAL", "Unable to get the Capture extension.");
                }
                captureDevice = cap.CaptureOpenDevice(deviceName, sampleRate, format, 4096);
                this.format = format;
                this.sampleRate = sampleRate;

            }
            public unsafe void Dispose() {
                ALContext alc = ALContext.GetApi();
                Pointer<Context> context = alc.CreateContext(device, null);
                alc.MakeContextCurrent(context);
                if (alc.TryGetExtension(device, out Capture cap)) {
                    throw new APIBackendException("OpenAL", "Unable to get the Capture extension.");
                }
                cap.CaptureCloseDevice(captureDevice);
                alc.CloseDevice(device);
            }
            /// <summary>
            /// Starts capturing the audio.
            /// </summary>
            public void Start() {
                ALContext alc = ALContext.GetApi();
                unsafe {
                    Pointer<Context> context = alc.CreateContext(device, null);
                    alc.MakeContextCurrent(context);
                    if (alc.TryGetExtension(device, out Capture cap)) {
                        throw new APIBackendException("OpenAL", "Unable to get the Capture extension.");
                    }
                    cap.CaptureStart(captureDevice);
                }
            }
            /// <summary>
            /// Stops capturing the audio and returns the sound.
            /// </summary>
            /// <returns>The captured sound.</returns>
            public Sound Stop() {
                ALContext alc = ALContext.GetApi();
                unsafe {
                    Pointer<Context> context = alc.CreateContext(device, null);
                    alc.MakeContextCurrent(context);
                    if (alc.TryGetExtension(device, out Capture cap)) {
                        throw new APIBackendException("OpenAL", "Unable to get the Capture extension.");
                    }
                    List<byte> buffer = [];
                    cap.CaptureStop(captureDevice);
                    int remainingSamples = cap.GetAvailableSamples(captureDevice);
                    if (remainingSamples > 0) {
                        cap.CaptureSamples();
                    }
                }
            }
        }
    }
}